### ISSUE TITLE
Bi-manual Canonicalization in SDP

### ISSUE THREAD

**Issue by JINXER000** [amazon-science/Spherical_Diffusion_Policy#4]
Thanks for your open-source code.  I had a question about the bi-manual relative action representation.

In the paper, you describe canonicalizing the state and action relative to each gripper for bi-manual setups. However, I noticed in the code (e.g., sdp_data.py) that the point cloud is explicitly shifted by subtracting the end-effector position. For the single-arm case, this makes sense, but I'm not sure how to apply this for the bi-manual case.

Should the point cloud be shifted twice (once per gripper)? Or is there another way you handle this for bi-manual inputs?

Besides, I noticed that the normalizer in the dual-arm case is not implemented, e.g., `sdp/dataset/robomimic_replay_point_cloud_gripper_equal_scale_dataset.py:185`.  Could you suggest the correct implementation? Thanks!

**Comment by ZXP-S-works** [comment]
Thanks for asking this question!

Yes, you can shifting point cloud twice. Meanwhile, a more efficient way is to change the "origin convolution" to be bi-manual convolution in the last layer (see Figure A5). In this case, you don't need to change the data normalizer.

Best,
xp

**Comment by ZXP-S-works** [comment]
FYI, pass the origin of the grippers (in the point cloud frame) to the adaptive_origin_pool layer: https://github.com/amazon-science/Spherical_Diffusion_Policy/blob/ea25aa5a0f16c1499448a05cd751bccec4b3e916/sdp/model/equiformer_v2/connectivity.py#L168

**Comment by JINXER000** [comment]
Thanks for your reply! So far I have tried shifting the point cloud twice. However, the result is worse than no canonization.  I guess one cause is that the point cloud observation is coupled with eef poses, and the translation invariance may not be effective. 
Could you please check my implementation below:
```python
            # A_T0 - e_T0 
            nactions[:, :, 0, :3] -= nobs['robot0_eef_pos'][:, -1:]
            # A_T1 - e_T1 
            nactions[:, :, 1, :3] -= nobs['robot1_eef_pos'][:, -1:]
            # O - e_T0 - e_T1
            nobs['point_cloud'][:, :, :, :3] -= (nobs['robot0_eef_pos'][:, None, -1:] + nobs['robot1_eef_pos'][:, None, -1:])
            # e_T0 - e_T0
            nobs['robot0_eef_pos'][:, -1:] -= nobs['robot0_eef_pos'][:, -1:]
            # e_T1 - e_T1
            nobs['robot1_eef_pos'][:, -1:] -= nobs['robot1_eef_pos'][:, -1:]
```

**Comment by ZXP-S-works** [comment]
Here, the point cloud is canonicalized twice, which is in a strange frame.

There are two options:
a. The complex one, canonicalize observation-action to each of the grippers (use origin in the adaptive_origin_pool).
b. The simple one, canonicalize observation-action to the center of the grippers (treat two grippers as "one gripper").

Perhaps you could try option b), which is more convenient to implement.

**Comment by JINXER000** [comment]
Thanks for your reply!
I tried to implement option (a) by revising equiformer_enc.py:

```
        ########### Downstream Block #############
        for n, block in enumerate(self.down_blocks):
            #### Downsampling ####
            # For the last pooling stage (AdaptiveOriginPool), optionally pass gripper origins (bi-manual)
            if n == len(self.down_blocks) - 1:
                # Build origins tensor of shape (batch_size, K, 3)
                # Use the latest proprio timestep as origin per gripper
                origins_list = [ee_pose_0[:, :, -1]]  # (b, 3)
                if ee_pose_1 is not None:
                    origins_list.append(ee_pose_1[:, :, -1])
                origion = None
                if len(origins_list) > 0:
                    # stack along K dimension if multiple grippers
                    origion = torch.stack(origins_list, dim=1)  # (b, K, 3)
                pool_graph = block['pool'](node_coord_src=node_coord, batch_src=batch, origion=origion)
            else:
                pool_graph = block['pool'](node_coord_src=node_coord, batch_src=batch)
            node_coord_dst, edge_src, edge_dst, degree, batch_dst, node_idx = pool_graph
```
I have trained the model with canonication=False. 
However, the performance is even worse than before. 
Could you provide an example implementation? Which one do you use during your real-world experiment? Thanks!
